[MM-23180] Updated URL regex and added new tests

willypuzzle opened the pull request:
#### Summary
The URL regular expression in the marked.js library has been updated to include full-width punctuation intervals for Unicode characters. This change improves the handling of URLs with special characters. 

Additionally, two new test files have been added: link_with_special_text.html and link_with_special_text.text, which are designed to test the improved URL parsing functionality with special text cases.

#### Ticket Link
Fixes: https://github.com/mattermost/mattermost/issues/13971

Related PR:
- https://github.com/mattermost/mattermost/pull/30084

willypuzzle committed:
Updated URL regex and version, added new tests

The URL regular expression in the marked.js library has been updated to include full-width punctuation intervals for Unicode characters. This change improves the handling of URLs with special characters. The package version has also been incremented from 0.3.6 to 0.3.7 reflecting this update.

Additionally, two new test files have been added: link_with_special_text.html and link_with_special_text.text, which are designed to test the improved URL parsing functionality with special text cases.

hmhealey commented on package.json:
```diff
@@ -2,7 +2,7 @@
   "description": "A markdown parser built for speed",
   "author": "Christopher Jeffrey",
-  "version": "0.3.6",
+  "version": "0.3.7",
```
Thanks for the PR! Could you revert this? We're not publishing our forked package, so I don't think we need to update this

willypuzzle committed:
Update package.json

willypuzzle commented:
@hmhealey I can import this package on mattermost?

hmhealey commented:
> @hmhealey how I can import this package on mattermost?

If you mean for temporary testing, you can use [`npm link`](https://docs.npmjs.com/cli/v9/commands/npm-link), but keep in mind that you might need to restart Webpack to get it to pick up additional changes you make after that.

If you mean for submission though, you can update the web app's `package.json` from `"marked": "github:mattermost/marked#2ef7f28cc7718e3f551c4ce9ea75fdd7580c2008",` to `"marked": "github:willypuzzle/marked#e1c22578e19f48ca3ab2e4349d2a9b4634850079",`. Once that's submitted, it'll build the web app against your branch here.

willypuzzle commented:
> > @hmhealey how I can import this package on mattermost?
> 
> If you mean for temporary testing, you can use [`npm link`](https://docs.npmjs.com/cli/v9/commands/npm-link), but keep in mind that you might need to restart Webpack to get it to pick up additional changes you make after that.
> 
> If you mean for submission though, you can update the web app's `package.json` from `"marked": "github:mattermost/marked#2ef7f28cc7718e3f551c4ce9ea75fdd7580c2008",` to `"marked": "github:willypuzzle/marked#e1c22578e19f48ca3ab2e4349d2a9b4634850079",`. Once that's submitted, it'll build the web app against your branch here.

This won't be merged? I would refer to that of mattermost in the code of mattermost's PR

hmhealey commented:
Sorry, I meant that for if you were wanting to test these changes during development and while we review them. Generally the way that I make changes to this code is:
1. Make the changes locally and use `npm link` to test them with my local web app
2. Submit two PRs: one here and one in the monorepo which temporarily updates Marked to the PR branch in the package.json and adds some web app tests
3. Review both PRs in parallel
4. Merge the Marked PR and then update the web app PR to use the final commit
5. Merge the web app PR

It's a bit clunky unfortunately, but it ensures that we can review and QA everything in the web app

willypuzzle commented:
> Sorry, I meant that for if you were wanting to test these changes during development and while we review them. Generally the way that I make changes to this code is:
> 
> 1. Make the changes locally and use `npm link` to test them with my local web app
> 2. Submit two PRs: one here and one in the monorepo which temporarily updates Marked to the PR branch in the package.json and adds some web app tests
> 3. Review both PRs in parallel
> 4. Merge the Marked PR and then update the web app PR to use the final commit
> 5. Merge the web app PR
> 
> It's a bit clunky unfortunately, but it ensures that we can review and QA everything in the web app

For monorepo do you mean my repo?

willypuzzle committed:
Updated inline latex rendering

Changed the way inline latex is rendered in the Renderer class. The change involves switching from template literals to single quotes for string encapsulation, ensuring proper escaping of text.

willypuzzle committed:
Added markdown parser functionality

Implemented a markdown parser with various functionalities including block-level grammar, normal block grammar, GFM block grammar, and GFM + Tables block grammar. The lexer has been set up to handle these different grammars based on the options provided. Also added preprocessing and lexing methods for handling source text input.

willypuzzle committed:
Merge remote-tracking branch 'origin/MM-23180-Added-Support-For-Full-Width-Punctuation-Exclusion' into MM-23180-Added-Support-For-Full-Width-Punctuation-Exclusion

willypuzzle commented:
@hmhealey I tried to put `"marked": "github:willypuzzle/marked#8dba39e0cee718d19dbb4cf5417d5155f6f7d48b"` in the `package.json` of `webapp/channels` directory and call npm install but in the `node_modules/marked` directory I see always the old version. Why?

willypuzzle commented:
You can see https://github.com/mattermost/mattermost/pull/30084

willypuzzle commented:
@hmhealey don't you think it's time to merge this PR?

M-ZubairAhmed commented on lib/marked.js:
```diff
@@ -1066,7 +1066,7 @@ Renderer.prototype.codespan = function(text) {
 Renderer.prototype.inlinelatex = function(text) {
   //Return a helper object that is replaced inside of mattermost.
-  return `<span data-inline-latex="${escape(text)}"></span>`;
+  return '<span data-inline-latex="${escape(text)}"></span>';
```
can you please explain this change from ` to '

willypuzzle replied:
When I tried to transpile the code, in order to build the build. The transpiler complained about the use of backtick.

M-ZubairAhmed replied:
Why did it not complain before?

willypuzzle replied:
I don't know, I think a lot of time passed since the last build.

M-ZubairAhmed replied:
> he transpiler complained about the use of backtick.

Can you please send the complain here

willypuzzle replied:
you can read it [here](https://pastebin.com/tTAThAEG)

hmhealey replied:
I'll pull this locally and see if I can sort out the issue since this change would definitely break inline latex support

hmhealey replied:
When I tried running `make`, I got an error that uglifyjs wasn't found, and it wasn't installed by `npm install`. I've never run that though, so it shouldn't be needed.

Could you revert this, the changes to `package-lock.json`, and the addition of the root `marked.js`?

willypuzzle replied:
Maybe you should install uglifyjs globally. Anyway, I revered the changes as requested

willypuzzle committed:
Revert "Updated inline latex rendering"

This reverts commit 544b845fdff1daffde2b6466925ae1a9a06b9f19.

willypuzzle committed:
Revert "Added markdown parser functionality"

This reverts commit 881e98fed8eef5838b9f1cca07aedf8db185a65a.
